lib/commons/aria/implicit-role.js

```diff
@@ -9,86 +9,25 @@ import lookupTable from './lookup-table';
  * @param {HTMLElement} node The node to test
  * @return {Mixed} Either the role or `null` if there is none
  */
-// TODO: axe 4.0 - rename to `getImplicitRole` so we can use
-// a variable named `implicitRole` when assigning the output of
-// the function
 function implicitRole(node) {
-	'use strict';
+	const vNode = getNodeFromTree(node);
 
-	/*
-	 * Filter function to reduce a list of roles to a valid list of roles for a nodetype
-	 */
-	const isValidImplicitRole = function(set, role) {
-		const validForNodeType = function(implicitNodeTypeSelector) {
-			// TODO: es-module-utils.matchesSelector
-			return axe.utils.matchesSelector(node, implicitNodeTypeSelector);
-		};
-
-		if (role.implicit && role.implicit.some(validForNodeType)) {
-			set.push(role.name);
-		}
-
-		return set;
-	};
-
-	/*
-	 * Score a set of roles and aria-attributes by its optimal score
-	 * E.g. [{score: 2, name: button}, {score: 1, name: main}]
-	 */
-	const sortRolesByOptimalAriaContext = function(roles, ariaAttributes) {
-		const getScore = function(role) {
-			const allowedAriaAttributes = allowedAttr(role);
-			return allowedAriaAttributes.reduce(function(score, attribute) {
-				return score + (ariaAttributes.indexOf(attribute) > -1 ? 1 : 0);
-			}, 0);
-		};
-
-		const scored = roles.map(function(role) {
-			return { score: getScore(role), name: role };
-		});
-
-		const sorted = scored.sort(function(scoredRoleA, scoredRoleB) {
-			return scoredRoleB.score - scoredRoleA.score;
-		});
-
-		return sorted.map(function(sortedRole) {
-			return sortedRole.name;
-		});
-	};
-
-	/*
-	 * Create a list of { name / implicit } role mappings to filter on
-	 */
-	const roles = Object.keys(lookupTable.role).map(function(role) {
-		const lookup = lookupTable.role[role];
-		return { name: role, implicit: lookup && lookup.implicit };
-	});
+	if (!vNode) {
+		return null;
```

Comment:
Maybe? Some of our tests don't include the full tree (<body> got here for some of our tests and wasn't included in the tree). But it could also be the element is excluded from the context when doing a parent/child lookup.

Reply:
What do you mean with "excluded from the context when doing a parent/child lookup"? I definitely don't think it's right to conclude that if we don't have it in the tree the role is null. Better to have it throw I think.